api/desecapi/views.py

```diff
@@ -247,34 +248,26 @@ def get_queryset(self):
 
     def get_object(self):
         queryset = self.filter_queryset(self.get_queryset())
-
-        rrset_type = self.kwargs['type']
-
-        if rrset_type in RRset.RESTRICTED_TYPES:
-            raise PermissionDenied("You cannot tinker with the %s RRset." % rrset_type)
-
-        obj = get_object_or_404(queryset, type=rrset_type, subname=self.kwargs['subname'])
-
+        obj = get_object_or_404(queryset, type=self.kwargs['type'], subname=self.kwargs['subname'])
         self.check_object_permissions(self.request, obj)
-
         return obj
 
     def update(self, request, *args, **kwargs):
-        if not isinstance(request.data, dict):
-            raise ValidationError({
-                api_settings.NON_FIELD_ERRORS_KEY: ['Invalid input, expected a JSON object.']
-            }, code='invalid')
-
         # Attach URL parameters (self.kwargs) to the request body object (request.data),
         # the latter having preference with both are given.
+        data = deepcopy(request.data)
         for k in ('type', 'subname'):
-            # This works because we exclusively use JSONParser which causes request.data to be
-            # a dict (and not an immutable QueryDict, as is the case for other parsers)
-            request.data[k] = request.data.pop(k, self.kwargs[k])
+            data[k] = request.data.pop(k, self.kwargs[k])
+
+        partial = kwargs.pop('partial', False)
+        instance = self.get_object()
+        serializer = self.get_serializer(instance, data=data, partial=partial)
+        serializer.is_valid(raise_exception=True)
 
         try:
             with PDNSChangeTracker():
-                response = super().update(request, *args, **kwargs)
+                self.perform_update(serializer)
```

Comment:
Wouldn't it be more appropriate to have the change tracker in `perform_update()`, and have the next line (`response` assignment) outside the `with` block?

Reply:
I moved all change trackers into the perform methods, c427edf9cdecd9340b859e54af8544c619614354